File: reference/rn-engine/src/games/sudoku/gameplay/analysis/techniqueModules/hiddenPair.ts
import { sudokuDigits } from '../../../types';
import { allHouseRefs, buildCandidateEliminationMove, getCellCandidatesByIndex, getHouseDigitMatches } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';
import type { SudokuCanonicalMove } from '../moves';

export const hiddenPairTechnique: SudokuTechniqueDispatcher = {
  technique: 'hidden-pair',
  tier: 'medium',
  findMove(state) {
    let best: SudokuCanonicalMove | null = null;
    let bestComplexity = Infinity;

    for (const { house, cells } of allHouseRefs) {
      const complexity = cells.filter((i) => state.board[i] === 0).length * 1.5;
      if (complexity >= bestComplexity) {
        continue;
      }

      const digitPositions = new Map<number, number[]>();
      sudokuDigits.forEach((digit) => {
        digitPositions.set(digit, getHouseDigitMatches(state, cells, digit));
      });

      let found = false;
      for (let leftIndex = 0; leftIndex < sudokuDigits.length && !found; leftIndex += 1) {
        for (let rightIndex = leftIndex + 1; rightIndex < sudokuDigits.length && !found; rightIndex += 1) {
          const leftDigit = sudokuDigits[leftIndex];
          const rightDigit = sudokuDigits[rightIndex];
          const leftPositions = digitPositions.get(leftDigit) ?? [];
          const rightPositions = digitPositions.get(rightDigit) ?? [];
          if (leftPositions.length !== 2 || rightPositions.length !== 2) {
            continue;
          }
          if (leftPositions[0] !== rightPositions[0] || leftPositions[1] !== rightPositions[1]) {
            continue;
          }

          const move = buildCandidateEliminationMove({
            technique: 'hidden-pair',
            eliminations: leftPositions.flatMap((index) => getCellCandidatesByIndex(state, index)
              .filter((digit) => digit !== leftDigit && digit !== rightDigit)
              .map((digit) => ({ index, digit }))),
            evidenceCells: leftPositions,
            houses: [house],
            complexity,
          });
          if (move) {
            best = move;
            bestComplexity = complexity;
            found = true;
          }
        }
      }
    }

    return best;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/HiddenPair.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object HiddenPairTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.HIDDEN_PAIR
    override val tier = Difficulty.MEDIUM

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        var best: SudokuCanonicalMove? = null
        var bestComplexity = Double.MAX_VALUE

        for (ref in allHouseRefs) {
            val complexity = ref.cells.count { state.board[it] == 0 } * 1.5
            if (complexity >= bestComplexity) continue

            val digitPositions = (1..9).associateWith { digit -> getHouseDigitMatches(state, ref.cells, digit) }
            var found = false
            for (leftIndex in 1..9) {
                if (found) break
                for (rightIndex in (leftIndex + 1)..9) {
                    val leftPositions = digitPositions.getValue(leftIndex)
                    val rightPositions = digitPositions.getValue(rightIndex)
                    if (leftPositions.size != 2 || rightPositions.size != 2) continue
                    if (leftPositions[0] != rightPositions[0] || leftPositions[1] != rightPositions[1]) continue

                    val eliminations = leftPositions.flatMap { index ->
                        getCellCandidatesByIndex(state, index).filter { it != leftIndex && it != rightIndex }.map { index to it }
                    }
                    val move = buildCandidateEliminationMove(SudokuTechnique.HIDDEN_PAIR, eliminations, leftPositions, listOf(ref.house), complexity.toInt())
                    if (move != null) {
                        best = move
                        bestComplexity = complexity
                        found = true
                        break
                    }
                }
            }
        }
        return best
    }
}
